src/Data/Deque/Mutable/Linear.hs

```diff
@@ -0,0 +1,182 @@
+  , addLeft
+  , addRight
+  , popLeft
+  , popRight
+  , map
+  -- * Consumption
+  , toList
+  )
+where
+
+import qualified Data.Array.Mutable.Linear as Array
+import Data.Unrestricted.Linear
+import Prelude.Linear hiding (length, map)
+import qualified Prelude
+import GHC.Stack
+
+
+-- # Types
+-------------------------------------------------------------------------------
+
+data Deque a where
+  Deque :: Int -> Maybe (Int,Int) -> !(Array.Array a) %1-> Deque a
+  -- This is: Deque size (Maybe (left ptr,right ptr)) array
+  --
+  -- The deque is represented as a slice from the left pointer
+  -- moving forwards (and wrapping the end) to the right pointer.
+  -- We are full when the right pointer is one less than the left pointer
+  -- modulo the array size. An empty deque has no left or right pointers
+  -- and we start a singleton deque with both pointers at 0.
+  --
+  -- Invariants:
+  -- > If empty, we store no pointers
+  -- > The first int is the size of the array
+  -- > If we store pointers l,r then [arr[l], ... arr[r]] is the deque
+
+data Dir = L | R -- A left or right direction
+
+-- # Internal Helpers
+-------------------------------------------------------------------------------
+
+-- Given a size, get the next index
+nextIx :: Int -> Int -> Int
+nextIx sz ix = (ix + 1) `mod` sz
+
+-- Given a size, get the previous index
+prevIx :: Int -> Int -> Int
+prevIx sz ix = (ix - 1 + sz) `mod` sz
+
+-- # Allocation
+-------------------------------------------------------------------------------
+
+-- | Run a computation of an empty Deque with a given size
+alloc :: HasCallStack => Int -> (Deque a %1-> Ur b) %1-> Ur b
+alloc k f = Array.alloc k err $ \arr -> f (Deque k Nothing arr) where
+  err = Prelude.error "Accessing error element of a collection!"
+
+-- | Run a computation on a Deque that is deterimined by the given the list
+-- where we treat the start and end of the list as the left and right pointers,
+-- with the total capacity as the length of the list.
+fromList :: [a] -> (Deque a %1-> Ur b) %1-> Ur b
+fromList xs f | l <- Prelude.length xs = Array.fromList xs $
+  \arr -> f (Deque l (Just (0,l-1)) arr)
+
+
+-- # Querying
+-------------------------------------------------------------------------------
+
+-- | The total capacity of the Deque
+size :: Deque a %1-> (Ur Int, Deque a)
+size (Deque sz ptrs arr) = (Ur sz, Deque sz ptrs arr)
+
+-- | The number of elements currently stored
+length :: Deque a %1-> (Ur Int, Deque a)
+length (Deque sz Nothing arr) = (Ur 0, Deque sz Nothing arr)
+length (Deque sz p@(Just (l,r)) arr)
+  | l <= r = (Ur (1 + r - l), Deque sz p arr)
+  | otherwise = (Ur (sz + 1 - l + r), Deque sz p arr)
+
+-- | We are full if the length equals the size
+isFull :: Deque a %1-> (Ur Bool, Deque a)
+isFull d =
+  length d & \(Ur len, Deque sz p arr) -> (Ur (len == sz), Deque sz p arr)
+
+peek :: HasCallStack => Dir -> Deque a %1-> (Ur a, Deque a)
+peek _ (Deque _ Nothing arr) = error "Peeking in empty deque." $ arr
```

Comment:
I would expect this function to return a `Maybe` (`(Ur (Maybe a), Deque a)`), I think it's a bit different than out-of-bounds array access; where when you're working with queues you usually don't expect them to have certain number of stuff inside and you are supposed to handle them being empty.

Reply:
I don't think so because it's analogous to `top` for stacks which should fail if the stack is empty.

Reply:
On second, thought, this ^ makes sense.